### PR TITLE
User can edit family member

### PR DESCRIPTION
**Pull request by xtrsyz**

### CODE REVIEW

**Comment by nafiesl** [comment]
Hello @xtrsyz, would you mind adding unit tests regarding these changes?
You may refer to the testing classes in this directory:

https://github.com/nafiesl/silsilah/tree/master/tests/Unit/Policies

I will review this PR after the testing has been added.

Thanks.

**Comment by xtrsyz** [comment]
sorry I don't understand how to add it.
and I don't have time to learn it again.
thanks.

**Comment by nafiesl** [comment]
Hi @xtrsyz, this project has been covered by automated testing. Every feature that has impacts on the database changes (create, edit or delete) must be covered by testing.

I can understand if you don't have time to write the tests. So I will leave this PR open until I or someone else can write tests for these changes.

Thanks for your idea and contribution. I appreciate your time working on this PR.

### Current Progress

1. Write logic that allows users to edit their own family members (new authorization)

### Next TODO

1. Write tests on [policy objects](https://github.com/nafiesl/silsilah/tree/master/tests/Unit/Policies) covering the user authorizations to edit family members.

**Comment by xtrsyz** [comment]
Hi @nafiesl, i have create test unit for this.
Thanks.

File: app/Policies/UserPolicy.php [hunk 1]
```diff
@@ -18,6 +18,17 @@ class UserPolicy
...
+
+        foreach ($user->husbands as $husband) {
+            if ($editableUser->id == $husband->id) return true;
+        }
+
+        foreach ($user->wifes as $wife) {
+            if ($editableUser->id == $wife->id) return true;
+        }
+
         return $editableUser->id == $user->id || $editableUser->manager_id == $user->id || is_system_admin($user);
```
Comment (nafiesl) [inline app/Policies/UserPolicy.php:32]: Kalau dibuat gini lebih readable ngga ya mas @xtrsyz?
```suggestion
        return $editableUser->id == $user->id
        || $editableUser->manager_id == $user->id
        || is_system_admin($user)
        || $editableUser->id == $user->father_id
        || $editableUser->id == $user->mother_id
        || $editableUser->father_id == $user->id
        || $editableUser->mother_id == $user->id
        || $user->husbands->pluck('id')->contains($editableUser->id)
        || $user->wifes->pluck('id')->contains($editableUser->id)
        ;
```

**Review by nafiesl — CHANGES_REQUESTED** [review]
Terima kasih mas @xtrsyz sudah kasih testingnya. Saya ada beberapa masukan terkait kodenya mas.

Kemudian ini ada kontribusi yang bagus juga dari mas @azharlihan, dimana testingnya dipisah2 sesuai konteks nya (siapa boleh meng-edit siapa).
https://github.com/nafiesl/silsilah/pull/105/files

File: tests/Unit/Policies/UserPolicyTest.php [hunk 2]
```diff
@@ -88,4 +88,68 @@ public function user_cannot_delete_their_own_data()
...
+        $otherUser = factory(User::class)->create();
+
+        $this->assertTrue($child->can('edit', $mother));
+        $this->assertFalse($otherUser->can('edit', $mother));
+    }
+
+    /** @test */
+    public function husband_can_edit_their_wifes_profile()
+    {
+        $couple = factory(Couple::class)->create();
```
Comment (nafiesl) [inline tests/Unit/Policies/UserPolicyTest.php:139]: mas @xtrsyz class `Couple` ini belum diimport.

screen_2023-08-01_008

Comment (xtrsyz) [reply tests/Unit/Policies/UserPolicyTest.php:140]: ok terimakasih, menambah ilmu baru.
image

**Review by nafiesl — APPROVED** [review]
Dah sip, thanks mas @xtrsyz

**Pull request merged**